lib/spack/spack/binary_distribution.py

```diff
@@ -0,0 +1,409 @@
+            elif spack.relocate.needs_text_relocation(filetype):
+                rel_path_name = os.path.relpath(path_name, spec.prefix)
+                text_to_relocate.append(rel_path_name)
+
+    # Create buildinfo data and write it to disk
+    buildinfo = {}
+    buildinfo['buildpath'] = spack.store.layout.root
+    buildinfo['relocate_textfiles'] = text_to_relocate
+    buildinfo['relocate_binaries'] = binary_to_relocate
+    filename = buildinfo_file_name(spec)
+    with open(filename, 'w') as outfile:
+        outfile.write(yaml.dump(buildinfo, default_flow_style=True))
+
+
+def install_directory_name(spec):
+    """
+    Return name of the install directory according to the convention
+    <os>-<architecture>/<compiler>/<package>-<version>-<dag_hash>/
+    """
+    return "%s/%s/%s/%s-%s-%s" % (spack.store.layout.root,
+                                  spack.architecture.sys_type(),
+                                  str(spec.compiler).replace("@", "-"),
+                                  spec.name, spec.version, spec.dag_hash())
+
+
+def tarball_directory_name(spec):
+    """
+    Return name of the tarball directory according to the convention
+    <os>-<architecture>/<compiler>/<package>-<version>/
+    """
+    return "%s/%s/%s-%s" % (spack.architecture.sys_type(),
+                            str(spec.compiler).replace("@", "-"),
+                            spec.name, spec.version)
+
+
+def tarball_name(spec, ext):
+    """
+    Return the name of the tarfile according to the convention
+    <os>-<architecture>-<package>-<dag_hash><ext>
+    """
+    return "%s-%s-%s-%s-%s%s" % (spack.architecture.sys_type(),
+                                 str(spec.compiler).replace("@", "-"),
+                                 spec.name,
+                                 spec.version,
+                                 spec.dag_hash(),
+                                 ext)
+
+
+def tarball_path_name(spec, ext):
+    """
+    Return the full path+name for a given spec according to the convention
+    <tarball_directory_name>/<tarball_name>
+    """
+    return os.path.join(tarball_directory_name(spec),
+                        tarball_name(spec, ext))
+
+
+def build_tarball(spec, outdir, force=False, rel=False, yes_to_all=False,
+                  key=None):
+    """
+    Build a tarball from given spec and put it into the directory structure
+    used at the mirror (following <tarball_directory_name>).
+    """
+    tarfile_name = tarball_name(spec, '.tar.gz')
+    tarfile_dir = join_path(outdir, "build_cache",
+                            tarball_directory_name(spec))
+    tarfile_path = join_path(tarfile_dir, tarfile_name)
+    mkdirp(tarfile_dir)
+    spackfile_path = os.path.join(
+        outdir, "build_cache", tarball_path_name(spec, '.spack'))
+    if os.path.exists(spackfile_path):
+        if force:
+            os.remove(spackfile_path)
+        else:
+            tty.warn("file exists, use -f to force overwrite: %s" %
```

Comment:
It looks like that's true, in https://docs.python.org/3/tutorial/errors.html 8.4:

> The sole argument to raise indicates the exception to be raised. This must be either an exception instance or an exception class (a class that derives from Exception).

So for example update:

```
class NoGpgException:
    pass
```

with

```
class NoGpgException(Exception):
    pass
```